File: codes/rust/496.下一个更大元素-i.rs
/*
 * @lc app=leetcode.cn id=496 lang=rust
 *
 * [496] 下一个更大元素 I
 */

// @lc code=start
use std::collections::HashMap;
impl Solution {
    pub fn next_greater_element(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        let mut ans = vec![-1; nums1.len()];

        // 定一个单调栈和一个哈希表，分别用来存储从栈顶到栈底单调递增的元素和数值的下一个更大元素的映射值
        let mut stack: Vec<i32> = Vec::new();
        let mut table: HashMap<i32, i32> = HashMap::new();
        // 逆序遍历nums2,求出nums2中每个元素的下一个更大元素
        for i in (0..nums2.len()).rev() {
            // 如果栈顶元素小于当前元素，就踢掉栈顶元素，知道找到下一个更大的元素或栈底
            while !stack.is_empty() && stack.last().unwrap() < &nums2[i] {
                stack.pop();
            }
            // 如果存在下一个更大元素，就将其存入哈希表
            if !stack.is_empty() {
                table.insert(nums2[i], stack.last().unwrap().to_owned());
            }
            // 将当前元素入栈
            stack.push(nums2[i]);
        }

        // 接着遍历nums1, 从哈希表中找到每个元素的下一个更大元素
        for i in 0..nums1.len() {
            if table.contains_key(&nums1[i]) {
                ans[i] = table.get(&nums1[i]).unwrap().to_owned();
            }
        }
        ans
    }
}
// @lc code=end

File: codes/java/496.下一个更大元素-i.java
class Solution {
	public int[] nextGreaterElement(int[] nums1, int[] nums2) {
		// 记录 nums2 中每个元素的下一个更大元素
		int[] greater = nextGreaterElement(nums2);
		// 转化成映射：元素 x -> x 的下一个最大元素
		HashMap<Integer, Integer> greaterMap = new HashMap<>();
		for (int i = 0; i < nums2.length; i++) {
			greaterMap.put(nums2[i], greater[i]);
		}
		// nums1 是 nums2 的子集，所以根据 greaterMap 可以得到结果
		int[] res = new int[nums1.length];
		for (int i = 0; i < nums1.length; i++) {
			res[i] = greaterMap.get(nums1[i]);
		}
		return res;
	}

	// 计算 nums 中每个元素的下一个更大元素
	int[] nextGreaterElement(int[] nums) {
		int n = nums.length;
		// 存放答案的数组
		int[] res = new int[n];
		Stack<Integer> s = new Stack<>();
		// 倒着往栈里放
		for (int i = n - 1; i >= 0; i--) {
			// 判定个子高矮
			while (!s.isEmpty() && s.peek() <= nums[i]) {
				// 矮个起开，反正也被挡着了。。。
				s.pop();
			}
			// nums[i] 身后的下一个更大元素
			res[i] = s.isEmpty() ? -1 : s.peek();
			s.push(nums[i]);
		}
		return res;
	}
}